## Область видимости функции

Если переменная определена внутри функции, снаружи обратиться к ней не получится. Это произойдёт из-за областей видимости. Область видимости — правила, которыми руководствуется движок, когда ищет переменную или функцию по имени (идентификатору).

### Правила поиска переменной

- 1. Если идентификатор создан внутри функции, обратиться к нему снаружи нельзя.
- 2. Если обратиться к идентификатору из функции, движок сначала поищет идентификатор в теле этой функции. Если не найдёт — начнёт искать снаружи.

### Глобальная область видимости

- Область видимости вне любых функций называется глобальной. Идентификаторы, определённые в ней, видны везде.
- Переменную называют глобальной, если она объявлена в глобальной области видимости, и локальной — если внутри функции:

### Вложенные области видимости

Области видимости «вкладываются» друг в друга:

```javascript
const a = 1;

function callMe() {
  const b = 2;

  // функция callMeToo вложена в CallMe
  function callMeToo() {
    const c = 3;

    console.log(a);
    console.log(b);
    console.log(c);
  }

  callMeToo();
}

callMe();

// 1
// 2
// 3
```

Если в теле функции callMeToo обратиться к идентификатору, движок сначала поищет её внутри callMeToo, если не найдёт — будет искать внутри функции callMe, и уже затем — в глобальной области видимости.

### Затенение идентификаторов

❗Внутренняя переменная затеняет внешнюю

Если в теле функции обратиться к идентификатору, движок JavaScript сначала ищет его в теле этой функции, и уже потом снаружи. Поэтому если и снаружи, и внутри есть одинаковые идентификаторы, движок остановится на внутреннем, а внешний — проигнорирует.

```javascript
const a = "Hello";

function callMe() {
  const a = "world";

  console.log(a);
}

callMe(); // "world"
```

```javascript
function callMeFromTheInside() {
  console.log("Hello");
}

function callMe() {
  function callMeFromTheInside() {
    console.log("world");
  }
  callMeFromTheInside();
}

callMe(); // "world"
```

Когда движок увидел идентификатор callMeFromTheInside внутри функции callMe, он остановил свои поиски. Внутренняя функция затенила внешнюю.
